crates/holochain_zome_types/src/element.rs

```diff
@@ -235,13 +237,55 @@ impl HashableContent for SignedHeader {
 }
 
 /// The header and the signature that signed it
-#[derive(Clone, Debug, Eq, Serialize, Deserialize)]
-#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
-pub struct SignedHeaderHashed {
-    /// The hashed but unsigned header.
-    header: HeaderHashed,
+pub type SignedHeaderHashed = SignedHashed<Header>;
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[cfg_attr(feature = "test_utils", derive(arbitrary::Arbitrary))]
+/// A type that can be hashed and signed equivalently to a [`Header`].
+/// For example it is possible to hash and sign a [`Create`] by
+/// serializing it in a way that produces the same bytes as if
+/// it was serialized as a [`Header::Create`].
+pub struct SignedHashed<T>
+where
+    T: HashableContent<HashType = holo_hash::hash_type::Header>,
+{
+    /// The hash of this header.
+    pub hash: HeaderHash,
+    /// The data that can be serialized, hashed and signed
+    /// as a [`Header`].
+    pub header: T,
     /// The signature of the header.
-    signature: Signature,
+    pub signature: Signature,
+}
+
+impl<T> SignedHashed<T>
+where
+    T: HashableContent<HashType = holo_hash::hash_type::Header>,
+{
```

Comment:
I see your intention was to make this only about headers, and if you want this to not be fully generic, then I think you should replace this type constraint with something more specific to only headers. My concern is just that if you build a tool that's nice & generic, devs should be able to just pick it up and use it and not worry about using it in the wrong way. If there is a wrong way to use it, put that in the types.

Reply:
I see now that this was an intentional choice to limit it to header hashes only. You're right that I misunderstood the intent of this PR.

I'm OK to approve this as-is. I also think there is nothing lost in letting this be fully generic, since it's most of the way there.